File: go/test/annotation_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/mixpanel-annotations-sdk/go"
	"github.com/voxgig-sdk/mixpanel-annotations-sdk/go/core"

	vs "github.com/voxgig-sdk/mixpanel-annotations-sdk/go/utility/struct"
)

func TestAnnotationEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Annotation(nil)
		if ent == nil {
			t.Fatal("expected non-nil AnnotationEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := annotationBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "annotation." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		annotationRef01Ent := client.Annotation(nil)
		annotationRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "annotation"}), "annotation_ref01"))
		annotationRef01Data["project_id"] = setup.idmap["project01"]

		annotationRef01DataResult, err := annotationRef01Ent.Create(annotationRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		annotationRef01Data = core.ToMapAny(entityData(annotationRef01DataResult))
		if annotationRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if annotationRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		annotationRef01DataUp0Up := map[string]any{
			"id": annotationRef01Data["id"],
			"project_id": setup.idmap["project_id"],
		}

		annotationRef01MarkdefUp0Name := "date"
		annotationRef01MarkdefUp0Value := fmt.Sprintf("Mark01-annotation_ref01_%d", setup.now)
		annotationRef01DataUp0Up[annotationRef01MarkdefUp0Name] = annotationRef01MarkdefUp0Value

		annotationRef01ResdataUp0Result, err := annotationRef01Ent.Update(annotationRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		annotationRef01ResdataUp0 := core.ToMapAny(entityData(annotationRef01ResdataUp0Result))
		if annotationRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if annotationRef01ResdataUp0["id"] != annotationRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if annotationRef01ResdataUp0[annotationRef01MarkdefUp0Name] != annotationRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", annotationRef01MarkdefUp0Name, annotationRef01ResdataUp0[annotationRef01MarkdefUp0Name])
		}

		// LOAD
		annotationRef01MatchDt0 := map[string]any{
			"id": annotationRef01Data["id"],
		}
		annotationRef01DataDt0Loaded, err := annotationRef01Ent.Load(annotationRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		annotationRef01DataDt0LoadResult := core.ToMapAny(entityData(annotationRef01DataDt0Loaded))
		if annotationRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if annotationRef01DataDt0LoadResult["id"] != annotationRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		annotationRef01MatchRm0 := map[string]any{
			"id": annotationRef01Data["id"],
		}
		_, err = annotationRef01Ent.Remove(annotationRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func annotationBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "annotation", "AnnotationTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read annotation test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse annotation test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"annotation01", "annotation02", "annotation03", "project01", "project02", "project03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID": idmap,
		"MIXPANEL_ANNOTATIONS_TEST_LIVE":      "FALSE",
		"MIXPANEL_ANNOTATIONS_TEST_EXPLAIN":   "FALSE",
		"MIXPANEL_ANNOTATIONS_APIKEY":         "",
		"MIXPANEL_ANNOTATIONS_SERVER_REGIONANDDOMAIN": "mixpanel",
	})

	idmapResolved := core.ToMapAny(env["MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add project_id alias for update test.
	if idmapResolved["project_id"] == nil {
		idmapResolved["project_id"] = idmapResolved["project01"]
	}

	if env["MIXPANEL_ANNOTATIONS_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["MIXPANEL_ANNOTATIONS_APIKEY"],
				"server": map[string]any{
					"regionAndDomain": env["MIXPANEL_ANNOTATIONS_SERVER_REGIONANDDOMAIN"],
				},
			},
			extraOpts,
		})
		client = sdk.NewMixpanelAnnotationsSDK(core.ToMapAny(mergedOpts))
	}

	live := env["MIXPANEL_ANNOTATIONS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["MIXPANEL_ANNOTATIONS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/AnnotationEntityTest.php
<?php
declare(strict_types=1);

// Annotation entity test

require_once __DIR__ . '/../mixpanelannotations_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class AnnotationEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = MixpanelAnnotationsSDK::test(null, null);
        $ent = $testsdk->Annotation(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = annotation_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "annotation." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $annotation_ref01_ent = $client->Annotation(null);
        $annotation_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.annotation"), "annotation_ref01"));
        $annotation_ref01_data["project_id"] = $setup["idmap"]["project01"];

        $annotation_ref01_data_result = $annotation_ref01_ent->create($annotation_ref01_data, null);
        $annotation_ref01_data = Helpers::to_map(is_object($annotation_ref01_data_result) && method_exists($annotation_ref01_data_result, 'data_get') ? $annotation_ref01_data_result->data_get() : $annotation_ref01_data_result);
        $this->assertNotNull($annotation_ref01_data);
        $this->assertNotNull($annotation_ref01_data["id"]);

        // UPDATE
        $annotation_ref01_data_up0_up = [
            "id" => $annotation_ref01_data["id"],
            "project_id" => $setup["idmap"]["project_id"],
        ];

        $annotation_ref01_markdef_up0_name = "date";
        $annotation_ref01_markdef_up0_value = "Mark01-annotation_ref01_" . $setup["now"];
        $annotation_ref01_data_up0_up[$annotation_ref01_markdef_up0_name] = $annotation_ref01_markdef_up0_value;

        $annotation_ref01_resdata_up0_result = $annotation_ref01_ent->update($annotation_ref01_data_up0_up, null);
        $annotation_ref01_resdata_up0 = Helpers::to_map(is_object($annotation_ref01_resdata_up0_result) && method_exists($annotation_ref01_resdata_up0_result, 'data_get') ? $annotation_ref01_resdata_up0_result->data_get() : $annotation_ref01_resdata_up0_result);
        $this->assertNotNull($annotation_ref01_resdata_up0);
        $this->assertEquals($annotation_ref01_resdata_up0["id"], $annotation_ref01_data_up0_up["id"]);
        $this->assertEquals($annotation_ref01_resdata_up0[$annotation_ref01_markdef_up0_name], $annotation_ref01_markdef_up0_value);

        // LOAD
        $annotation_ref01_match_dt0 = [
            "id" => $annotation_ref01_data["id"],
        ];
        $annotation_ref01_data_dt0_loaded = $annotation_ref01_ent->load($annotation_ref01_match_dt0, null);
        $annotation_ref01_data_dt0_load_result = Helpers::to_map(is_object($annotation_ref01_data_dt0_loaded) && method_exists($annotation_ref01_data_dt0_loaded, 'data_get') ? $annotation_ref01_data_dt0_loaded->data_get() : $annotation_ref01_data_dt0_loaded);
        $this->assertNotNull($annotation_ref01_data_dt0_load_result);
        $this->assertEquals($annotation_ref01_data_dt0_load_result["id"], $annotation_ref01_data["id"]);

        // REMOVE
        $annotation_ref01_match_rm0 = [
            "id" => $annotation_ref01_data["id"],
        ];
        $annotation_ref01_ent->remove($annotation_ref01_match_rm0, null);

    }
}

function annotation_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/annotation/AnnotationTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = MixpanelAnnotationsSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["annotation01", "annotation02", "annotation03", "project01", "project02", "project03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID" => $idmap,
        "MIXPANEL_ANNOTATIONS_TEST_LIVE" => "FALSE",
        "MIXPANEL_ANNOTATIONS_TEST_EXPLAIN" => "FALSE",
        "MIXPANEL_ANNOTATIONS_APIKEY" => "",
        "MIXPANEL_ANNOTATIONS_SERVER_REGIONANDDOMAIN" => 'mixpanel',
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["MIXPANEL_ANNOTATIONS_TEST_ANNOTATION_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["project_id"])) {
        $idmap_resolved["project_id"] = $idmap_resolved["project01"];
    }

    if ($env["MIXPANEL_ANNOTATIONS_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
                "apikey" => $env["MIXPANEL_ANNOTATIONS_APIKEY"],
                "server" => [
                    "regionAndDomain" => $env["MIXPANEL_ANNOTATIONS_SERVER_REGIONANDDOMAIN"],
                ],
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        // "?? []" because merge legitimately answers with a stdClass when every
        // contributing entry is an EMPTY map - an SDK with no apikey and no
        // server variables generates an empty middle entry, so that is the
        // common case, not the edge one. to_map returns null for a non-array by
        // design, and the constructor takes a non-nullable array, so without the
        // fallback every such SDK died on "must be of type array, null given"
        // the moment live mode was switched on. Offline mode never reaches this
        // branch, which is why the offline suite stayed green.
        $client = new MixpanelAnnotationsSDK(Helpers::to_map($merged_opts) ?? []);
    }

    $live = $env["MIXPANEL_ANNOTATIONS_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["MIXPANEL_ANNOTATIONS_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
